src/Data/List/Properties.agda

```diff
@@ -740,6 +714,39 @@ map-concatMap f g xs = begin
   concatMap (map f ∘′ g) xs
     ∎
 
+------------------------------------------------------------------------
+-- catMaybes
+
+catMaybes-concatMap : catMaybes {A = A} ≗ concatMap fromMaybe
+catMaybes-concatMap []             = refl
+catMaybes-concatMap (just x  ∷ xs) = cong (x ∷_) (catMaybes-concatMap xs)
+catMaybes-concatMap (nothing ∷ xs) = catMaybes-concatMap xs
+
+length-catMaybes : ∀ xs → length (catMaybes {A = A} xs) ≤ length xs
+length-catMaybes []             = ≤-refl
+length-catMaybes (just x  ∷ xs) = s≤s (length-catMaybes xs)
```

Comment:
Personally I think it's clearer to not do 2-level pattern matching but to delegate the job to `maybe`, i.e.
```agda
length-catMaybes (x  ∷ xs) = let l = length-catMaybes xs) in maybe (\ _ → s≤s l) (m≤n⇒m≤1+n l) x
```
which more cleanly separates the list-based reasoning from the maybe-based reasoning.

(In my own code, I'd got further and use higher-order instead of `let` but I've been convinced that this is a step too far for stdlib)

Reply:
Also: `where` instead of `let` cf. your own #2369 ?

Reply:
>  Personally I think it's clearer to not do 2-level pattern matching but to delegate the job to maybe

Your suggested change doesn't typecheck though: using eliminators in dependent situations often requires that you explicitly provide the 'motive' (in this case generalising over `mx`). Indeed that's what people used to do before proper dependent pattern matching came along:
```agda
length-catMaybes : ∀ xs → length (catMaybes {A = A} xs) ≤ length xs
length-catMaybes []        = ≤-refl
length-catMaybes (mx ∷ xs) = let IH = length-catMaybes xs in
  maybe {B = λ mx → length (catMaybes (mx ∷ xs)) ≤ length (mx ∷ xs)}
        (λ _ → s≤s IH)
        (m≤n⇒m≤1+n IH)
        mx
```
It gets even worse in more complicated scenarios like `catMaybes-↭` where you need to introduce more variables, etc.

A quick `grep` through the codebase reveals no other explicit motives for `maybe`: are you sure you want this change (and most importantly, maintaining this code)?

Btw, `length-catMaybes` was already there :rofl:

Reply:
I was definitely overly optimistic regarding `maybe` being able to infer the motive. As it can't, then this 3-line proof is indeed the better one.